pkg/apis/apm/v1/apmserver_types.go

```diff
@@ -112,37 +118,118 @@ func (as *ApmServer) IsMarkedForDeletion() bool {
 	return !as.DeletionTimestamp.IsZero()
 }
 
-func (as *ApmServer) ElasticsearchRef() commonv1.ObjectSelector {
-	return as.Spec.ElasticsearchRef
-}
-
 func (as *ApmServer) SecureSettings() []commonv1.SecretSource {
 	return as.Spec.SecureSettings
 }
 
-func (as *ApmServer) AssociationConf() *commonv1.AssociationConf {
-	return as.assocConf
-}
-
 func (as *ApmServer) ServiceAccountName() string {
 	return as.Spec.ServiceAccountName
 }
 
-func (as *ApmServer) SetAssociationConf(assocConf *commonv1.AssociationConf) {
-	as.assocConf = assocConf
-}
-
 // EffectiveVersion returns the version reported by APM server. For development builds APM server does not use the SNAPSHOT suffix.
 func (as *ApmServer) EffectiveVersion() string {
 	return strings.TrimSuffix(as.Spec.Version, "-SNAPSHOT")
 }
 
-func (as *ApmServer) AssociationStatus() commonv1.AssociationStatus {
-	return as.Status.Association
+func (as *ApmServer) GetAssociations() []commonv1.Association {
+	return []commonv1.Association{
+		&ApmEsAssociation{ApmServer: as},
+		&ApmKibanaAssociation{ApmServer: as},
+	}
+}
+
+// ApmEsAssociation helps to manage the APMServer / Elasticsearch association
+type ApmEsAssociation struct {
+	*ApmServer
+}
+
+var _ commonv1.Association = &ApmEsAssociation{}
+
+func (aes *ApmEsAssociation) Associated() commonv1.Associated {
+	if aes == nil {
+		return nil
+	}
+	if aes.ApmServer == nil {
+		aes.ApmServer = &ApmServer{}
+	}
+	return aes.ApmServer
+}
+
+func (aes *ApmEsAssociation) AnnotationName() string {
+	return "association.k8s.elastic.co/es-conf"
+}
+
+func (aes *ApmEsAssociation) AssociatedServiceType() string {
+	return "elasticsearch"
+}
+
+func (aes *ApmEsAssociation) AssociationRef() commonv1.ObjectSelector {
+	return aes.Spec.ElasticsearchRef
```

Comment:
We seem to call `AssociationRef().WithDefaultNamespace()` in many places, which seems a bit error-prone. Would it be better to have this function return the ref with the correct namespace (also in other associations)?